File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>	
#include <vector>
#include <algorithm>
#include <cstring>
#include <string>
using namespace std;

int n, arr[26][26], cnt; const int dx[] = { 0,0,-1,1 }, dy[] = { 1,-1,0,0 };
bool one[26][26];	vector<int> set_size;

void dfs(int y, int x) {
	for (int i = 0; i < 4; i++) {
		int d_y = y + dy[i], d_x = x + dx[i];
		if (d_y < 0 || d_y >= n || d_x < 0 || d_x >= n)	continue;
		if (one[d_y][d_x]) {
			one[d_y][d_x] = false;
			cnt++;
			dfs(d_y, d_x);
		}
	}
}

int main() {
	ios_base::sync_with_stdio(false);	cin.tie(NULL);
	cin >> n;
	memset(one, false, sizeof(one));
	for (int i = 0; i < n; i++) {
		string a;	
		cin >> a;
		for (int j = 0; j < n; j++) {
			arr[i][j] = a[j] - '0';
			if (arr[i][j] == 1)
				one[i][j] = true;
		}
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (one[i][j]) {
				cnt = 1;
				one[i][j] = false;
				dfs(i, j);
				set_size.push_back(cnt);
			}
		}
	}
	cout << set_size.size() << endl;
	sort(set_size.begin(), set_size.end());
	for (vector<int>::iterator iter = set_size.begin(); iter != set_size.end(); iter++) {
		int temp = *iter;
		cout << temp << endl;
	}
	return 0;
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;

public class Main{
    private static int dx[] = {0,0,-1,1};
    private static int dy[] = {1,-1,0,0};
    private static int[] aparts = new int[25*25];
    private static int n;
    private static int cnt = 0;
    private static boolean[][] visited = new boolean[25][25];
    private static int[][] map = new int[25][25];
    
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        
        map = new int[n][n];
        visited = new boolean[n][n];
        
        for(int i = 0; i < n; i++){
            String temp = sc.next();
            for(int j = 0; j < n; j++){
                map[i][j] = temp.charAt(j) - '0';
            }
        }
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(map[i][j] == 1 && !visited[i][j]){
                    cnt++;
                    dfs(i,j);
                }
            }
        }
        Arrays.sort(aparts);
        System.out.println(cnt);
        
        for(int i = 0; i < aparts.length; i++){
            if(aparts[i] != 0){
                System.out.println(aparts[i]);
            }
        }
    }
    
    private static void dfs(int x, int y){
        visited[x][y] = true;
        aparts[cnt]++;
        for(int i = 0; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(nx >= 0 && ny >= 0 && nx < n && ny < n){
                if(map[nx][ny] == 1 && !visited[nx][ny]){
                    dfs(nx,ny);
                }
            }
        }
    }
}